src/compiler/transformers/module/module.ts

```diff
@@ -671,28 +707,53 @@ namespace ts {
 
         function visitExpressionStatement(node: ExpressionStatement): VisitResult<Statement> {
             const original = getOriginalNode(node);
-            if (original.kind === SyntaxKind.EnumDeclaration
-                && hasModifier(original, ModifierFlags.Export)) {
-                return visitExpressionStatementForEnumDeclaration(node, <EnumDeclaration>original);
+            const origKind = original.kind;
+
+            if (origKind === SyntaxKind.EnumDeclaration || origKind === SyntaxKind.ModuleDeclaration) {
+                return visitExpressionStatementForEnumOrNamespaceDeclaration(node, <EnumDeclaration | ModuleDeclaration>original);
+            }
+            else if (origKind === SyntaxKind.ClassDeclaration) {
+                // The decorated assignment for a class name will potentially need to be transformed.
+                const classDecl = original as ClassDeclaration;
+                if (classDecl.name) {
+                    const statements = [node];
+                    // Avoid emitting a default because that will typically be taken care of for us.
+                    if (hasModifier(classDecl, ModifierFlags.Export) && !hasModifier(classDecl, ModifierFlags.Default)) {
+                        addExportMemberAssignment(statements, classDecl)
+                    }
+
+                    addExportMemberAssignments(statements, classDecl.name);
+
+                    if (statements.length > 1) {
+                        Debug.assert(!!classDecl.decorators, "Expression statements should only have an export member assignment when decorated.")
+                    }
+                    return statements;
+                }
             }
 
             return node;
         }
 
-        function visitExpressionStatementForEnumDeclaration(node: ExpressionStatement, original: EnumDeclaration): VisitResult<Statement> {
-            if (isFirstDeclarationOfKind(original, SyntaxKind.EnumDeclaration)) {
-                const statements: Statement[] = [node];
-                addVarForExportedEnumDeclaration(statements, original);
-                return statements;
+        function visitExpressionStatementForEnumOrNamespaceDeclaration(node: ExpressionStatement, original: EnumDeclaration | ModuleDeclaration): VisitResult<Statement> {
+            const statements: Statement[] = [node];
+
+            // Preserve old behavior for enums in which a variable statement is emitted after the body itself.
+            if (hasModifier(original, ModifierFlags.Export) &&
+                original.kind === SyntaxKind.EnumDeclaration &&
+                isFirstDeclarationOfKind(original, SyntaxKind.EnumDeclaration)) {
+                addVarForExportedEnumOrNamespaceDeclaration(statements, original);
+            }
+            else {

```

Comment:
It shouldn't be else  but irrespecitive
eg. 
export enum e {
}
export { e as e2}